File: csharp/Solitaire/DeckTests/CardTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Silnith.Game.Deck.Tests
{
    [TestClass]
    public class CardTests
    {
        [TestMethod]
        public void TestCard()
        {
            Card _ = new(Value.Eight, Suit.Heart);
        }

        [TestMethod]
        public void TestGetValue()
        {
            Card card = new(Value.Eight, Suit.Heart);

            Assert.AreEqual(Value.Eight, card.Value);
        }

        [TestMethod]
        public void TestGetSuit()
        {
            Card card = new(Value.Eight, Suit.Heart);

            Assert.AreEqual(Suit.Heart, card.Suit);
        }

        [TestMethod]
        public void TestEquals()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Eight, Suit.Heart);

            Assert.AreEqual(card1, card2);
        }

        [TestMethod]
        public void TestHashCode()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Eight, Suit.Heart);

            Assert.AreEqual(card1.GetHashCode(), card2.GetHashCode());
        }

        [TestMethod]
        public void TestEqualsDifferentValue()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Seven, Suit.Heart);

            Assert.AreNotEqual(card1, card2);
        }

        [TestMethod]
        public void TestHashCodeDifferentValue()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Seven, Suit.Heart);

            Assert.AreNotEqual(card1.GetHashCode(), card2.GetHashCode());
        }

        [TestMethod]
        public void TestEqualsDifferentSuit()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Eight, Suit.Diamond);

            Assert.AreNotEqual(card1, card2);
        }

        [TestMethod]
        public void TestHashCodeDifferentSuit()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Eight, Suit.Diamond);

            Assert.AreNotEqual(card1.GetHashCode(), card2.GetHashCode());
        }

        [TestMethod]
        public void TestEqualsDifferentValueAndSuit()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Seven, Suit.Diamond);

            Assert.AreNotEqual(card1, card2);
        }

        [TestMethod]
        public void TestHashCodeDifferentValueAndSuit()
        {
            Card card1 = new(Value.Eight, Suit.Heart);
            Card card2 = new(Value.Seven, Suit.Diamond);

            Assert.AreNotEqual(card1.GetHashCode(), card2.GetHashCode());
        }

        [TestMethod]
        public void TestEqualsDifferentType()
        {
            Card card = new(Value.Eight, Suit.Heart);

            Assert.AreNotEqual<object>(card, "Eight of Hearts");
        }

        [TestMethod]
        public void TestToString()
        {
            Card card = new(Value.Eight, Suit.Heart);

            Assert.AreEqual("Eight of Hearts", card.ToString());
        }
    }
}

File: cpp/Solitaire/DeckTests/card_tests.cpp
#include "CppUnitTest.h"

#include <silnith/game/deck/card.h>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

using namespace silnith::game::deck;

using namespace std::literals::string_literals;

template<>
std::wstring Microsoft::VisualStudio::CppUnitTestFramework::ToString<>(value const& value)
{
	return to_wstring(value);
}

template<>
std::wstring Microsoft::VisualStudio::CppUnitTestFramework::ToString<>(suit const& suit)
{
	return to_wstring(suit);
}

template<>
std::wstring Microsoft::VisualStudio::CppUnitTestFramework::ToString<>(card const& card)
{
	return to_wstring(card);
}

namespace DeckTests
{
	TEST_CLASS(CardTests)
	{
	public:
		TEST_METHOD(TestGetValue)
		{
			card card{ value::eight, suit::heart };

			Assert::AreEqual(value::eight, card.get_value());
		}

		TEST_METHOD(TestGetSuit)
		{
			card card{ value::eight, suit::heart };

			Assert::AreEqual(suit::heart, card.get_suit());
		}

		TEST_METHOD(TestEquals)
		{
			card card1{ value::eight, suit::heart };
			card card2{ value::eight, suit::heart };

			Assert::AreEqual(card1, card2);
		}

		TEST_METHOD(TestEqualsDifferentValue)
		{
			card card1{ value::eight, suit::heart };
			card card2{ value::seven, suit::heart };

			Assert::AreNotEqual(card1, card2);
		}

		TEST_METHOD(TestEqualsDifferentSuit)
		{
			card card1{ value::eight, suit::heart };
			card card2{ value::eight, suit::diamond };

			Assert::AreNotEqual(card1, card2);
		}

		TEST_METHOD(TestEqualsDifferentValueAndSuit)
		{
			card card1{ value::eight, suit::heart };
			card card2{ value::seven, suit::diamond };

			Assert::AreNotEqual(card1, card2);
		}

		TEST_METHOD(TestToString)
		{
			card card{ value::ace, suit::spade };

			Assert::AreEqual("ace of spades"s, to_string(card));
		}
	};
}
